In [21]:
import pandas as pd
import pylab as plt
import numpy as np




In [22]:
df = pd.read_csv("G:\Data science\Capstone_insurance\Purchase.csv")

In [23]:
import xgboost

In [24]:
print(df)

        QuoteNumber Original_Quote_Date  QuoteConversion_Flag Field6  Field7  \
0                 1          2013-08-16                     0      B      23   
1                 2          2014-04-22                     0      F       7   
2                 4          2014-08-25                     0      F       7   
3                 6          2013-04-15                     0      J      10   
4                 8          2014-01-25                     0      E      23   
5                12          2014-01-18                     0      E      14   
6                13          2013-11-01                     0      J      23   
7                14          2014-05-14                     0      B      25   
8                18          2013-06-19                     0      J      10   
9                19          2013-06-18                     0      F      11   
10               20          2013-09-24                     0      B      23   
11               22          2013-06-11 

In [44]:
print(df.isnull().sum())

QuoteConversion_Flag    0
Field6                  0
Field7                  0
Field8                  0
Field9                  0
Field10                 0
Field11                 0
Field12                 0
CoverageField1A         0
CoverageField1B         0
CoverageField2A         0
CoverageField2B         0
CoverageField3A         0
CoverageField3B         0
CoverageField4A         0
CoverageField4B         0
CoverageField5A         0
CoverageField5B         0
CoverageField6A         0
CoverageField6B         0
CoverageField8          0
CoverageField9          0
CoverageField11A        0
CoverageField11B        0
SalesField1A            0
SalesField1B            0
SalesField2A            0
SalesField2B            0
SalesField3             0
SalesField4             0
                       ..
GeographicField51A      0
GeographicField51B      0
GeographicField52A      0
GeographicField52B      0
GeographicField53A      0
GeographicField53B      0
GeographicField54A      0
GeographicFi

In [18]:
import pandas as pd
from IPython.core.display import HTML

#df=pd.read_pickle('Data1')
#display(HTML(df.to_html()))

In [15]:
df = df.set_index('Original_Quote_Date')

df.index = df.index.to_datetime()


AttributeError: 'Index' object has no attribute 'to_datetime'

In [25]:
df['Date']=pd.to_datetime(pd.Series(df['Original_Quote_Date']))

In [26]:
df = df.drop(["Original_Quote_Date","QuoteNumber"],axis=1)

In [27]:
df['weekday'] = df['Date'].dt.dayofweek
df['weekyear'] = df['Date'].dt.weekofyear
df['Year'] = df['Date'].dt.year
df['Month'] = df['Date'].dt.month

In [28]:
df=df.drop(["Date"], axis=1)

In [29]:
df.head()

QuoteConversion_Flag Field6  Field7  Field8  Field9 Field10  Field11  \
0                     0      B      23  0.9403  0.0006     965   1.0200   
1                     0      F       7  1.0006  0.0040     548   1.2433   
2                     0      F       7  1.0006  0.0040     548   1.2433   
3                     0      J      10  0.9769  0.0004   1,165   1.2665   
4                     0      E      23  0.9472  0.0006   1,487   1.3045   

  Field12  CoverageField1A  CoverageField1B  ...    GeographicField61A  \
0       N               17               23  ...                    -1   
1       N                6                8  ...                    -1   
2       N                7               12  ...                    -1   
3       N                3                2  ...                    -1   
4       N                8               13  ...                    -1   

   GeographicField61B  GeographicField62A  GeographicField62B  \
0                  18                  -1                  10   
1                  17                  -1                  20   
2                  11                  -1                   8   
3                   9                  -1                  21   
4                  11                  -1                  12   

   GeographicField63  GeographicField64  weekday  weekyear  Year  Month  
0                  N                 CA        4        33  2013      8  
1                  N                 NJ        1        17  2014      4  
2                  N                 NJ        0        35  2014      8  
3                  N                 TX        0        16  2013      4  
4                  N                 IL        5         4  2014      1  

[5 rows x 301 columns]

In [40]:
import sklearn
from sklearn import preprocessing

In [30]:
print(df)

        QuoteConversion_Flag Field6  Field7  Field8  Field9 Field10  Field11  \
0                          0      B      23  0.9403  0.0006     965   1.0200   
1                          0      F       7  1.0006  0.0040     548   1.2433   
2                          0      F       7  1.0006  0.0040     548   1.2433   
3                          0      J      10  0.9769  0.0004   1,165   1.2665   
4                          0      E      23  0.9472  0.0006   1,487   1.3045   
5                          0      E      14  0.9472  0.0006   1,487   1.3045   
6                          0      J      23  0.9258  0.0004   1,165   1.2665   
7                          0      B      25  0.9153  0.0007     935   1.0200   
8                          0      J      10  0.9691  0.0004   1,165   1.2665   
9                          0      F      11  0.9919  0.0038     564   1.1886   
10                         0      B      23  0.9403  0.0006     965   1.0200   
11                         0      B     

In [42]:
a=[]
for columns in df.columns:
    if df[columns].dtype=="object":
        #a.append(columns)
        lbl=preprocessing.LabelEncoder()
        lbl.fit(list(df[columns].values))
        df[columns]=lbl.transform(list(df[columns].values))
#print(a)

In [68]:
df=df.fillna(-99)

In [70]:
df.isnull().sum().sum()

0

In [61]:
#df.columns[df.isna().any()].tolist()
#df['PersonalField84']

['PersonalField84', 'PropertyField29']

In [71]:
df.head()

QuoteConversion_Flag  Field6  Field7  Field8  Field9  Field10  Field11  \
0                     0       1      23  0.9403  0.0006        7   1.0200   
1                     0       5       7  1.0006  0.0040        4   1.2433   
2                     0       5       7  1.0006  0.0040        4   1.2433   
3                     0       6      10  0.9769  0.0004        1   1.2665   
4                     0       4      23  0.9472  0.0006        3   1.3045   

   Field12  CoverageField1A  CoverageField1B  ...    GeographicField61A  \
0        0               17               23  ...                    -1   
1        0                6                8  ...                    -1   
2        0                7               12  ...                    -1   
3        0                3                2  ...                    -1   
4        0                8               13  ...                    -1   

   GeographicField61B  GeographicField62A  GeographicField62B  \
0                  18                  -1                  10   
1                  17                  -1                  20   
2                  11                  -1                   8   
3                   9                  -1                  21   
4                  11                  -1                  12   

   GeographicField63  GeographicField64  weekday  weekyear  Year  Month  
0                  1                  0        4        33  2013      8  
1                  1                  2        1        17  2014      4  
2                  1                  2        0        35  2014      8  
3                  1                  3        0        16  2013      4  
4                  1                  1        5         4  2014      1  

[5 rows x 301 columns]

In [74]:
X=df.copy()
X=X.drop(['QuoteConversion_Flag'],axis=1)

In [75]:
y=df['QuoteConversion_Flag']

In [110]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y, 
                                                    test_size=0.4,
                                                    random_state=0)

In [124]:
from sklearn.model_selection import StratifiedKFold
from sklearn.model_selection import cross_val_score
kfold=StratifiedKFold(n_splits=2, random_state=7)


In [125]:
result=cross_val_score(clf,X_train,y_train,cv=kfold)

C:\Users\sam\Anaconda3\lib\site-packages\sklearn\preprocessing\label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
C:\Users\sam\Anaconda3\lib\site-packages\sklearn\preprocessing\label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:


In [127]:
result.mean()

0.9242702200948082

In [128]:
result.std()

0.00020405283584057354

In [111]:
import xgboost as xgb

In [112]:
#clf=xgb.XGBClassifier(n_estimators=100, learning_rate=0.01,nthread=-1,max_depth=12,subsample=0.6,colsample_bytree=0.6,
                      #silent=False,Verbose=10)

In [113]:
#clf=xgb.XGBClassifier(objective = 'binary:logistic', eta=0.1,n_estimators=100,silent=False,verbose=10)
#clf=xgb.XGBClassifier(objective = 'binary:logistic', eta=0.1,n_estimators=100,max_depth=10,silent=False,verbose=10)
#clf=xgb.XGBClassifier(objective = 'binary:logistic', eta=0.1,n_estimators=100,max_depth=10,alpha=1,silent=False,verbose=10)
clf=xgb.XGBClassifier(objective = 'binary:logistic', eta=0.1,n_estimators=100,max_depth=7,colsample_bytree=0.8,
                      alpha=1,gamma=0.1,subsample=0.8,silent=False,verbose=10)


In [114]:
clf.fit(X_train, y_train)

XGBClassifier(alpha=1, base_score=0.5, booster='gbtree', colsample_bylevel=1,
       colsample_bytree=0.8, eta=0.1, gamma=0.1, learning_rate=0.1,
       max_delta_step=0, max_depth=7, min_child_weight=1, missing=None,
       n_estimators=100, n_jobs=1, nthread=None,
       objective='binary:logistic', random_state=0, reg_alpha=0,
       reg_lambda=1, scale_pos_weight=1, seed=None, silent=False,
       subsample=0.8, verbose=10)

In [115]:
clf_xgb=clf.predict(X_test)

C:\Users\sam\Anaconda3\lib\site-packages\sklearn\preprocessing\label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:


In [116]:
from sklearn import metrics
from sklearn.tree import DecisionTreeClassifier
from sklearn.tree import export_graphviz
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix, recall_score, precision_score, accuracy_score
from sklearn.cross_validation import cross_val_score

In [117]:
print(confusion_matrix(y_test, clf_xgb))
print(classification_report(y_test,clf_xgb))

[[83326  1428]
 [ 6232 13316]]
             precision    recall  f1-score   support

          0       0.93      0.98      0.96     84754
          1       0.90      0.68      0.78     19548

avg / total       0.93      0.93      0.92    104302



# Precison - 0.95; recall - 0.68; Fl score - 0.74
#with params: eta =0.1, n_estimators=100,verbose=10

# Precison - 0.95; recall - 0.68; Fl score - 0.78
#with params: eta =0.1, n_estimators=100,maxdepth=10,verbose=10

# Precison - 0.95; recall - 0.68; Fl score - 0.78
#with params: eta =0.1, n_estimators=100,alpha=1,maxdepth=10,verbose=10